File: test-client/test-register.mjs
/**
 * Test MoltMart registration flow with ERC-8004 identity.
 * 
 * New architecture:
 * 1. Mint ERC-8004 identity (if needed) - x402 payment
 * 2. Get challenge message
 * 3. Sign challenge with wallet
 * 4. Register (FREE) with signature
 */

import { wrapFetchWithPaymentFromConfig } from '@x402/fetch';
import { ExactEvmScheme } from '@x402/evm/exact/client';
import { privateKeyToAccount } from 'viem/accounts';
import fs from 'fs';
import path from 'path';

// Load private key
const keyPath = path.join(process.env.HOME, '.openclaw/workspace/.kyro-wallet-key');
const privateKey = fs.readFileSync(keyPath, 'utf8').trim();

const account = privateKeyToAccount(privateKey);
console.log(`Wallet: ${account.address}`);

// Wrap fetch with x402 payment support for Base
const fetchWithPayment = wrapFetchWithPaymentFromConfig(fetch, {
  schemes: [
    {
      network: 'eip155:8453', // Base mainnet
      client: new ExactEvmScheme(account),
    },
  ],
});

const API_URL = 'https://api.moltmart.app';

async function checkERC8004() {
  console.log('\n1. Checking ERC-8004 status...');
  
  const response = await fetch(`${API_URL}/agents/8004/${account.address}`);
  const data = await response.json();
  
  if (data.verified || data.has_8004) {
    console.log(`   ✅ Already has ERC-8004 identity`);
    return true;
  }
  
  console.log(`   ❌ No ERC-8004 identity found`);
  return false;
}

async function mintIdentity() {
  console.log('\n2. Minting ERC-8004 identity ($0.05 USDC)...');
  
  const response = await fetchWithPayment(`${API_URL}/identity/mint`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ wallet_address: account.address }),
  });
  
  const data = await response.json();
  
  console.log(`   Response status: ${response.status}`);
  console.log(`   Response data: ${JSON.stringify(data, null, 2)}`);
  
  if (response.ok && data.success) {
    console.log(`   ✅ Minted! Agent ID: ${data.agent_id}`);
    console.log(`   TX: ${data.scan_url}`);
    return true;
  } else {
    console.log(`   ❌ Minting failed: ${data.error || response.status}`);
    if (data.reason) console.log(`   Reason: ${data.reason}`);
    if (data.detail) console.log(`   Detail: ${data.detail}`);
    return false;
  }
}

async function getChallenge() {
  console.log('\n3. Getting challenge message...');
  
  const response = await fetch(`${API_URL}/agents/challenge`);
  const data = await response.json();
  
  console.log(`   Challenge: "${data.challenge}"`);
  return data.challenge;
}

async function signChallenge(challenge) {
  console.log('\n4. Signing challenge...');
  
  const signature = await account.signMessage({ message: challenge });
  console.log(`   Signature: ${signature.slice(0, 20)}...`);
  return signature;
}

async function register(signature) {
  console.log('\n5. Registering (FREE)...');
  
  const registrationData = {
    name: 'Kyro',
    wallet_address: account.address,
    signature: signature,
    description: 'AI agent building MoltMart. Testing new registration flow.',
    moltx_handle: 'Kyro',
    github_handle: 'kyro-agent',
  };
  
  const response = await fetch(`${API_URL}/agents/register`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(registrationData),
  });
  
  console.log(`   Status: ${response.status}`);
  
  if (response.ok) {
    const result = await response.json();
    console.log('\n✅ REGISTERED!');
    console.log(`   API Key: ${result.api_key}`);
    console.log(`   Agent ID: ${result.id}`);
    if (result.erc8004) {
      console.log(`   ERC-8004 ID: ${result.erc8004.agent_id}`);
    }
    
    // Save API key
    const apiKeyPath = path.join(process.env.HOME, '.openclaw/workspace/.moltmart-api-key');
    fs.writeFileSync(apiKeyPath, result.api_key, { mode: 0o600 });
    console.log(`\n📝 API key saved to ${apiKeyPath}`);
    
    return result;
  } else {
    const error = await response.json();
    console.log(`   ❌ Registration failed: ${error.detail}`);
    return null;
  }
}

async function main() {
  console.log('=== MoltMart Registration Test ===\n');
  
  // Step 1: Check if already has ERC-8004
  const hasIdentity = await checkERC8004();
  
  // Step 2: Mint if needed
  if (!hasIdentity) {
    const minted = await mintIdentity();
    if (!minted) {
      console.log('\n❌ Cannot proceed without ERC-8004 identity');
      process.exit(1);
    }
  }
  
  // Step 3: Get challenge
  const challenge = await getChallenge();
  
  // Step 4: Sign challenge
  const signature = await signChallenge(challenge);
  
  // Step 5: Register
  await register(signature);
}

main().catch(console.error);

File: test-register.py
#!/usr/bin/env python3
"""
Test x402 registration flow with ERC-8004 identity minting.
Uses my signing wallet to pay and register.
"""

import os
import json
import httpx
import time
from eth_account import Account
from eth_account.messages import encode_typed_data

# Load private key
KEY_PATH = os.path.expanduser("~/.openclaw/workspace/.kyro-wallet-key")
with open(KEY_PATH) as f:
    PRIVATE_KEY = f.read().strip()

account = Account.from_key(PRIVATE_KEY)
print(f"Wallet: {account.address}")

# MoltMart API
API_URL = "https://api.moltmart.app"

# USDC on Base
USDC_ADDRESS = "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913"
USDC_DECIMALS = 6


def create_x402_payment(payment_requirements: dict) -> dict:
    """Create and sign an x402 payment payload."""
    
    # Extract payment details
    pay_to = payment_requirements["payTo"]
    amount = int(payment_requirements["maxAmountRequired"])
    network = payment_requirements["network"]
    resource = payment_requirements["resource"]
    
    # x402 exact scheme uses EIP-712 typed data
    # Based on x402 spec
    nonce = str(int(time.time() * 1000))  # millisecond timestamp as nonce
    
    # The domain and types for EIP-712 signing
    # This matches the x402 exact scheme specification
    domain = {
        "name": "x402",
        "version": "1",
        "chainId": 8453,  # Base mainnet
    }
    
    types = {
        "EIP712Domain": [
            {"name": "name", "type": "string"},
            {"name": "version", "type": "string"},
            {"name": "chainId", "type": "uint256"},
        ],
        "PaymentAuthorization": [
            {"name": "from", "type": "address"},
            {"name": "to", "type": "address"},
            {"name": "amount", "type": "uint256"},
            {"name": "token", "type": "address"},
            {"name": "nonce", "type": "string"},
            {"name": "validUntil", "type": "uint256"},
        ],
    }
    
    valid_until = int(time.time()) + 300  # 5 minutes
    
    message = {
        "from": account.address,
        "to": pay_to,
        "amount": amount,
        "token": USDC_ADDRESS,
        "nonce": nonce,
        "validUntil": valid_until,
    }
    
    # Sign typed data
    signable = encode_typed_data(
        domain_data=domain,
        message_types=types,
        message_data=message,
    )
    signed = account.sign_message(signable)
    
    # Build payment payload
    payload = {
        "x402Version": 1,
        "scheme": "exact",
        "network": network,
        "payload": {
            "signature": signed.signature.hex(),
            "authorization": {
                "from": account.address,
                "to": pay_to,
                "amount": str(amount),
                "token": USDC_ADDRESS,
                "nonce": nonce,
                "validUntil": valid_until,
            },
        },
    }
    
    return payload


def register():
    """Register on MoltMart with x402 payment."""
    
    print("\n1. Calling POST /agents/register...")
    
    registration_data = {
        "name": "Kyro",
        "wallet_address": account.address,
        "description": "AI agent building MoltMart. Day 5.",
        "moltx_handle": "Kyro",
        "github_handle": "kyro-agent",
    }
    
    import base64
    
    with httpx.Client(timeout=60.0) as client:
        # First call - expect 402
        response = client.post(
            f"{API_URL}/agents/register",
            json=registration_data,
        )
        
        print(f"   Status: {response.status_code}")
        
        if response.status_code != 402:
            print(f"   Unexpected response: {response.text}")
            return
        
        # Parse 402 response from header (x402 v2 format)
        payment_header = response.headers.get("payment-required")
        if not payment_header:
            print("   No payment-required header!")
            return
        
        payment_info = json.loads(base64.b64decode(payment_header).decode())
        print(f"   Payment required: {json.dumps(payment_info, indent=2)}")
        
        if "accepts" not in payment_info or not payment_info["accepts"]:
            print("   No payment options!")
            return
        
        payment_req = payment_info["accepts"][0]
        print(f"\n2. Creating x402 payment...")
        print(f"   Amount: {int(payment_req['maxAmountRequired']) / 10**6} USDC")
        print(f"   Pay to: {payment_req['payTo']}")
        
        # Create signed payment
        payment_payload = create_x402_payment(payment_req)
        print(f"   Signed payment created")
        
        # Encode as base64
        import base64
        payment_header = base64.b64encode(
            json.dumps(payment_payload).encode()
        ).decode()
        
        print(f"\n3. Retrying with X-Payment header...")
        
        # Retry with payment
        response2 = client.post(
            f"{API_URL}/agents/register",
            json=registration_data,
            headers={"X-Payment": payment_header},
        )
        
        print(f"   Status: {response2.status_code}")
        
        if response2.status_code == 200:
            result = response2.json()
            print(f"\n✅ REGISTERED!")
            print(f"   API Key: {result.get('api_key')}")
            print(f"   Agent ID: {result.get('id')}")
            if result.get('erc8004'):
                print(f"   ERC-8004 ID: {result['erc8004'].get('agent_id')}")
                print(f"   Scan URL: {result['erc8004'].get('scan_url')}")
        else:
            print(f"   Error: {response2.text}")


if __name__ == "__main__":
    register()
